openmls/tests/book_code.rs

```diff
@@ -1663,3 +1663,176 @@ fn commit_builder() {
     // ANCHOR_END: alice_adds_bob_with_commit_builder
     _ = (mls_message_out, welcome, group_info)
 }
+
+#[openmls_test]
+fn not_join_group() {
+    let group_id = GroupId::from_slice(b"Test Group");
+
+    let alice_provider = &Provider::default();
+    let bob_provider = &Provider::default();
+
+    // Generate credentials with keys
+    let (alice_credential, alice_signature_keys) = generate_credential(
+        "Alice".into(),
+        ciphersuite.signature_algorithm(),
+        alice_provider,
+    );
+
+    let (bob_credential, bob_signature_keys) = generate_credential(
+        "Bob".into(),
+        ciphersuite.signature_algorithm(),
+        bob_provider,
+    );
+
+    // Define the MlsGroup configuration
+    let mls_group_create_config = MlsGroupCreateConfig::builder()
+        .ciphersuite(ciphersuite)
+        .use_ratchet_tree_extension(true) // NOTE: important
+        .build();
+
+    // === Alice creates a group ===
+    let mut alice_group = MlsGroup::new_with_group_id(
+        alice_provider,
+        &alice_signature_keys,
+        &mls_group_create_config,
+        group_id.clone(),
+        alice_credential,
+    )
+    .expect("An unexpected error occurred.");
+
+    // Generate KeyPackages
+    let bob_key_package = generate_key_package(
+        ciphersuite,
+        bob_credential.clone(),
+        Extensions::default(),
+        bob_provider,
+        &bob_signature_keys,
+    );
+
+    // === Alice adds Bob ===
+    let (_mls_message_out, welcome, _group_info) = alice_group
+        .add_members(
+            alice_provider,
+            &alice_signature_keys,
+            &[bob_key_package.key_package().clone()],
+        )
+        .expect("Could not add members.");
+
+    let welcome: MlsMessageIn = welcome.into();
+
+    let group_context: Option<GroupContext> =
+        bob_provider.storage().group_context(&group_id).unwrap();
+    assert!(group_context.is_none());
+    let confirmation_tag: Option<ConfirmationTag> =
+        bob_provider.storage().confirmation_tag(&group_id).unwrap();
+    assert!(confirmation_tag.is_none());
+
+    let state_before = GroupStorageState::from_storage(bob_provider.storage(), &group_id);
+    //ANCHOR: not_join_group_welcome
+    let welcome = match welcome.extract() {
+        MlsMessageBodyIn::Welcome(welcome) => welcome,
+        _ => unimplemented!("Handle other message types"),
+    };
+    //ANCHOR_END: not_join_group_welcome
+
+    //ANCHOR: not_join_group_welcome_inspect
+    for _secret in welcome.secrets() {
```

Comment:
I can't think of a good reason why the secrets should be investigated. I actually wonder why this is exposed.